feat(providers): add Azure DevOps provider

nawok opened the pull request:
## ☕️ Reasoning

This pull request adds the new `AzureDevOpsProvider` which allows authenticating with [Azure DevOps](https://dev.azure.com/) and getting tokens for calling the APIs.

## 🧢 Checklist

- [x] Documentation
- [ ] Tests
- [x] Ready to be merged

## 🎫 Affected issues

Implements: #4688

## 📌 Resources

- [Contributing guidelines](./CONTRIBUTING.md)
- [Code of conduct](./CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

nawok commented on docs/docs/providers/azure-devops.md:
````diff
@@ -0,0 +1,107 @@
+
+## Example
+
+```js
````
Thanks @ndom91, I added the titles according to the [Code blocks](https://docusaurus.io/docs/markdown-features/code-blocks#code-title) manual.

nawok commented:
Dear @ThangHuuVu, @lluia, please let me know if you want me to set up an application in Azure DevOps so you can verify that the provider works.

nawok commented:
I decided to improve my pull request slightly (hope you find these additions helpful).

- Deployed NextAuth Example with Azure DevOps provider: https://next-auth-azure-devops.vercel.app
- The provider is copied “as-is” from this pull request and the full diff and usage can be found here: https://github.com/nawok/next-auth-example/compare/main...nawok:next-auth-example:provider/azure-devops?expand=1

ndom91 commented:
LGTM too. I can confirm the preview URL login with AzureDevops worked for me :+1:

nawok commented:
I also added a little fix for eslint’s incorrect suggestion to use `??` inside `if` condition.

```
next-auth:lint: next-auth/packages/next-auth/src/utils/detect-host.ts
next-auth:lint:   4:26  error  Prefer using nullish coalescing operator (`??`) instead of a logical or (`||`), as it is a safer operator  @typescript-eslint/prefer-nullish-coalescing
```

thejessewinton commented:
+1. Would love to see this integrated.

nawok commented:
After updating this PR I noticed that the `legacy` label was applied. What should I update/improve to modernize it?

ThangHuuVu commented:
Hi @nawok to get rid of the legacy label, could you help move the change under: https://github.com/nextauthjs/next-auth/tree/main/packages/core/src/providers

And the docs do not need to be written manually anymore - make sure to add proper JSDoc comments, for example https://github.com/nextauthjs/next-auth/blob/main/packages/core/src/providers/auth0.ts

nawok commented:
Is incorporating the docs as JSDoc a recommended way?

I found a separate Markdown has a better DX because of auto formatting and local preview. Please, have a look. There are some strange types generated on the bottom of the page and I cannot figure out what I am missing to make the page look nicer.

ndom91 commented:
> Is incorporating the docs as JSDoc a recommended way?
> 
> I found a separate Markdown has a better DX because of auto formatting and local preview. Please, have a look. There are some strange types generated on the bottom of the page and I cannot figure out what I am missing to make the page look nicer.

Yeah so we use the jsdoc comments to autogenerate the API reference docs. That way the editor documentation shown on hover, i.e. types, and the reference docs, are based on the same information. Of course it comes with some downsides, but having nice vscode built-in docs and maintaining only 1 set seems worth it.

In regard to the strange types at the bottom of the page, what exactly are you referring to?

nawok commented:
Hi Nico, yeah, makes sense. Helps to keep the docs in sync as well.

I was talking about this [`options`](https://auth-docs-9j870k11e-authjs.vercel.app/reference/core/functions/providers_azure_devops.default#parameters)’s type signature. I just checked other providers and it seems like the type «explodes» if you add an extra field to it. I just thought I did something wrong but now I see that I’m not alone—and this is just the way it is.

ThangHuuVu reviewed (commented):
hi @nawok sorry for the delayed review, I checked your branch but I am getting an error: 
```
"response" is not a conform Token Endpoint response
```

Digging in the [doc](https://app.vssps.visualstudio.com/oauth2/authorize?response_type=Assertion&scope=vso.profile&client_id=888B2758-D252-4617-A15D-C04C3B55EC32&redirect_uri=https%3A%2F%2Flocalhost%3A3000%2Fapi%2Fauth%2Fcallback%2Fazure-devops&code_challenge=oQgWKdkLpyFbQ7wXLWO4Cyfq35c5v59IOmGvEURpx9s&code_challenge_method=S256) & the response, I see:
```
Error: unsupported_grant_type. grant_type must be the ietf jwt-bearer type, refresh_token, or client_credentials'
``` 

Am I missing something?

nawok commented:
***@***.***, I will have a look today. Thank you for starting the review.

Best regards,
Pavel
On Aug 27, 2023 at 10:28 +0200, nextauthjs/next-auth ***@***.***>, wrote:
>
> Am

nawok commented:
Hey @ThangHuuVu, I rebased my changes on top of the latest `main`, simplified it a bit, and I also redeployed the [example app](https://next-auth-azure-devops.vercel.app) and here is its [diff](https://github.com/nextauthjs/next-auth-example/compare/main...nawok:nextauthjs---next-auth-example:provider/azure-devops?expand=1) with the stock version.

Can you give it another try, please?

ThangHuuVu commented on packages/core/src/providers/azure-devops.ts:
```diff
@@ -0,0 +1,174 @@
+      params: { response_type: 'Assertion', scope }
+    },
+
+    token: {
```
The example app works, could you link the example repo? I find it a bit weird because in the code, you're missing `url` for both `token` and `userinfo` config, so Auth.js is complaining: 
```

next-auth-app:dev: [auth][error][InvalidEndpoints]: Provider "azure-devops" is missing both `issuer` and `token` endpoint config. At least one of them is required.. Read more at https://errors.authjs.dev#invalidendpoints
next-auth-app:dev:     at assertConfig (webpack-internal:///(middleware)/../../../packages/core/lib/assert.js:76:24)
next-auth-app:dev:     at Auth (webpack-internal:///(middleware)/../../../packages/core/index.js:82:89)
next-auth-app:dev:     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

Could you add back `url` to `token` and `userinfo`?

ThangHuuVu reviewed (commented):
Could you also add the following changes: 
- .github/ISSUE_TEMPLATE/2_bug_provider.yml
- Add an image to docs/static/img/providers/

Everything else LGTM!

nawok commented:
Regarding the images, I was hesitating because of [Microsofts guidelines](https://www.microsoft.com/en-us/legal/intellectualproperty/trademarks) regarding their assets. I can create a proper icon of matching style and size if you insist.

nawok committed:
feat(providers): add Azure DevOps provider

nawok commented:
Re-added the endpoint URLs and updated the bug template dropdown options.

ThangHuuVu commented:
Woa, that is some strict guidelines. It should not be a blocker, though.

ThangHuuVu commented:
merged, thanks again @nawok!
